### PR TITLE
[3] access grant data

### LINKED ISSUE #3: Access research grant data
- [x] Set up data access configs
- [x] Use `data_getters` to access GtR data. 
- [x] Write a small script that builds the project table with the relevant fields.
- [x] Write and document functions and tests

### LINKED ISSUE #16: make install errors
First time usage I get the following three errors on `make install`. I can fix the first two, but the third requires some deeper digging

## 1: `Makefile:14: .env: No such file or directory`

Which can be remedied by creating `.env`, but probably should have some logic creates if it doesn't exist (i.e. `touch .env`)

## 2: `innovation_sweet_spots setup command: 'install_requires` ... `Parse error at "'+https:/'": Expected stringEnd`

After fixing the above I run into

```
make install
conda env create -q -n innovation_sweet_spots -f environment.yaml
Collecting package metadata (repodata.json): ...working... done
Solving environment: ...working... done
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
/Library/Developer/CommandLineTools/usr/bin/make -s pip-install
    ERROR: Command errored out with exit status 1:
     command: /Users/jklinger/anaconda3/envs/innovation_sweet_spots/bin/python -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/Users/jklinger/Nesta/innovation_sweet_spots/setup.py'"'"'; __file__='"'"'/Users/jklinger/Nesta/innovation_sweet_spots/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /private/var/folders/zc/xq14ps8175s62kfvt2_2r649xlpw4k/T/pip-pip-egg-info-4nsh16pl
         cwd: /Users/jklinger/Nesta/innovation_sweet_spots/
    Complete output (1 lines):
    error in innovation_sweet_spots setup command: 'install_requires' must be a string or list of strings containing valid project/version requirement specifiers; Parse error at "'+https:/'": Expected stringEnd
    ----------------------------------------
WARNING: Discarding file:///Users/jklinger/Nesta/innovation_sweet_spots. Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
make[1]: *** [pip-install] Error 1
make: *** [conda-create] Error 2
```

which requires changing `git+https://github.com/nestauk/data_getters.git` to `data_getters @ git+https://github.com/nestauk/data_getters.git` in `requirements.txt`

## 3. `make install` Hangs (recursive lookup?)

After the above fixes, the install instead hangs indefinitely (> 10 mins). I killed the command, which gives the following stack trace, indicating that there is a recursive lookup in the `pip-install` section of the Makefile

```
Traceback (most recent call last):
  File "/Users/jklinger/anaconda3/envs/innovation_sweet_spots/lib/python3.7/contextlib.py", line 509, in __exit__
    if cb(*exc_details):
  File "/Users/jklinger/anaconda3/envs/innovation_sweet_spots/lib/python3.7/contextlib.py", line 377, in _exit_wrapper
    return cm_exit(cm, exc_type, exc, tb)
  File "/Users/jklinger/anaconda3/envs/innovation_sweet_spots/lib/python3.7/site-packages/pip/_internal/utils/temp_dir.py", line 165, in __exit__
    self.cleanup()
  File "/Users/jklinger/anaconda3/envs/innovation_sweet_spots/lib/python3.7/site-packages/pip/_internal/utils/temp_dir.py", line 184, in cleanup
    rmtree(self._path)
  File "/Users/jklinger/anaconda3/envs/innovation_sweet_spots/lib/python3.7/site-packages/pip/_vendor/tenacity/__init__.py", line 339, in wrapped_f
    return self(f, *args, **kw)
  File "/Users/jklinger/anaconda3/envs/innovation_sweet_spots/lib/python3.7/site-packages/pip/_vendor/tenacity/__init__.py", line 430, in __call__
    do = self.iter(retry_state=retry_state)
  File "/Users/jklinger/anaconda3/envs/innovation_sweet_spots/lib/python3.7/site-packages/pip/_vendor/tenacity/__init__.py", line 367, in iter
    return fut.result()
  File "/Users/jklinger/anaconda3/envs/innovation_sweet_spots/lib/python3.7/concurrent/futures/_base.py", line 428, in result
    return self.__get_result()
  File "/Users/jklinger/anaconda3/envs/innovation_sweet_spots/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/Users/jklinger/anaconda3/envs/innovation_sweet_spots/lib/python3.7/site-packages/pip/_vendor/tenacity/__init__.py", line 433, in __call__
    result = fn(*args, **kwargs)
  File "/Users/jklinger/anaconda3/envs/innovation_sweet_spots/lib/python3.7/site-packages/pip/_internal/utils/misc.py", line 135, in rmtree
    shutil.rmtree(dir, ignore_errors=ignore_errors, onerror=rmtree_errorhandler)
  File "/Users/jklinger/anaconda3/envs/innovation_sweet_spots/lib/python3.7/shutil.py", line 496, in rmtree
    os.rmdir(path)
KeyboardInterrupt
make[1]: *** [pip-install] Error 1
make: *** [conda-create] Interrupt: 2
```

### PR DESCRIPTION
**Pull request by beingkk**
Closes #3; #16 

Added a function in `innovation_sweet_spots.getters.inputs` that basically wraps `data_getters.gtr.build_projects()` for easy fetching of GTR projects from Nesta database and storing them locally.

The intention is to eventually add other functions in `innovation_sweet_spots.getters.inputs` that will fetch any other required datasets. Then, the input dataset can be initialised (or updated) by simply running `python inputs.py` from the `getters` folder.

Tests for the getter functions can be found in `innovation_sweet_spots.tests.test_getters`

---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

### CODE REVIEW

**Comment by jaklinger** [comment]
Pre-review comment: Raised an issue for a problem setting up the repo according to the instructions in https://github.com/nestauk/innovation_sweet_spots/issues/16

File: innovation_sweet_spots/config/base.yaml
Comment (jaklinger) [inline innovation_sweet_spots/config/base.yaml]: suggest to use `git-crypt` so that this can be versioned (could be useful if we change database, which we will do in the next few months)

Also suggest to practice this on a blank text file before applying to the config file, although the file isn't *that* sensitive

File: README.md
Comment (jaklinger) [inline README.md:26]: I guess that under the current setup that a pre-instruction is to change the database config file in `base.yaml` (just a comment, I haven't checked if that's the case yet!)

File: innovation_sweet_spots/getters/inputs.py
Comment (jaklinger) [inline innovation_sweet_spots/getters/inputs.py]: PEP guidance is to capitalise globals

```suggestion
GTR_PATH = f"{PROJECT_DIR}/inputs/data/gtr_projects.csv"
```

File: innovation_sweet_spots/getters/inputs.py
Comment (jaklinger) [inline innovation_sweet_spots/getters/inputs.py]: Suggest not to re-download if a flag `use_cached=False` or the file doesn't already exist. I'd recommend that `use_cached=True`  by default, but that's a design decision for you

```suggestion
    if not (use_cached or os.path.exists(fpath)):
        projects = build_projects(
            config_path=config["database_config_path"],
            chunksize=5000,
            table_wildcards=["gtr_projects"],
            desired_fields=fields,
        )
        with open(fpath, "w") as f:
            json.dump(projects, f)
        logging.info(f"Collected {len(projects)} GTR projects and stored them in {fpath}")
    df = pd.read_json(fpath, orient='records')
    return df
```

File: innovation_sweet_spots/getters/inputs.py
Comment (jaklinger) [inline innovation_sweet_spots/getters/inputs.py]: Depends on your analysis, but it's not clear yet that `pandas` adds anything, so perhaps passing the raw `projects` object is more transparent at this point rather than converting to `df`

File: innovation_sweet_spots/tests/test_getters.py
Comment (jaklinger) [inline innovation_sweet_spots/tests/test_getters.py]:
```suggestion
        assert type(df) is DataFrame
```

File: innovation_sweet_spots/tests/test_getters.py
Comment (jaklinger) [inline innovation_sweet_spots/tests/test_getters.py]: this line won't get called if the test fails higher up - your options are either to wrap everything in a `try except` or to use `NamedTemporaryFile`

File: innovation_sweet_spots/tests/test_getters.py
Comment (jaklinger) [inline innovation_sweet_spots/tests/test_getters.py]: Line not required if using `pytest`

**Review by jaklinger — CHANGES_REQUESTED** [review]
Thanks for that - a few points for consideration (also see #16). It might be good at this stage to setup Travis integration as well, to run `pytest`.

Happy to discuss any details!

File: innovation_sweet_spots/getters/inputs.py
Comment (beingkk) [inline innovation_sweet_spots/getters/inputs.py]:
```suggestion
    return projects
```

File: innovation_sweet_spots/tests/test_getters.py
Comment (beingkk) [inline innovation_sweet_spots/tests/test_getters.py]:
```suggestion
```

File: innovation_sweet_spots/getters/inputs.py
Comment (beingkk) [reply innovation_sweet_spots/getters/inputs.py]: Sorry @jaklinger, I'm a bit confused about the `use_cached` flag. Should it be like another input variable to the get_gtr_projects() ? 

If `use_cache=True`, should it check the "cache" somewhere, or is it meant to simply say not to download anything?

Comment (beingkk) [reply innovation_sweet_spots/getters/inputs.py]: Oh, do I need to use the `@lru_cache` here?

Comment (jaklinger) [reply innovation_sweet_spots/getters/inputs.py]: exactly, so by cache i just mean that you already have a local copy (i.e. `os.path.exists(fpath)`) and so add `use_cached=True` to the kwargs for `get_gtr_projects` to avoid re-downloading the data every time

File: innovation_sweet_spots/tests/test_getters.py
Comment (beingkk) [reply innovation_sweet_spots/tests/test_getters.py]: Thanks, I didn't know about `NamedTemporaryFile` - a really nice suggestion!

File: README.md
Comment (beingkk) [reply README.md:26]: I took up your suggestion to use git-crypt, and add the encrypted config file to the repo for versioning purposes

File: innovation_sweet_spots/config/base.yaml
Comment (beingkk) [reply innovation_sweet_spots/config/base.yaml]: Thanks, that's what I did - also checked that the file is indeed encrypted.

File: innovation_sweet_spots/getters/inputs.py
Comment (beingkk) [reply innovation_sweet_spots/getters/inputs.py]: I adjusted the code to return the raw `projects` and to dump the projects as json file. Then I ran into issue with non-serializable types in `projects` (specifically, `pandas._libs.tslibs.nattype.NaTType` and `pandas._libs.tslibs.timestamps.Timestamp`), which I've circumvented using `json.dump(projects, f, default=datetime.datetime.isoformat)` and documented the implications in the docstring.

**Comment by beingkk** [comment]
Hi, thanks a lot for all the good suggestions! I've made the changes trying to reflect the suggestions, and you should also be able to set up the project now (please set it up from this branch, i.e. `3_access_grant_data`; it will not work from `dev` until it's merged).

P.S. ping me for the git-crypt key

**Comment by beingkk** [comment]
Good point about Travis! - I will look into setting it up (I've done it once before, for a repo from my own account)

File: innovation_sweet_spots/getters/inputs.py
Comment (jaklinger) [inline innovation_sweet_spots/getters/inputs.py]: You can drop some hard-coding here and refer back to the global, whilst also converting the second to lines to a safe one-liner

```suggestion
    data_folder = Path(GTR_PATH).parent
    data_folder.mkdir(parents=True, exist_ok=True)
```

**Comment by jaklinger** [comment]
Thanks for that - just before running could you drop the crypt-key in the `nesta-production-config` bucket? Thanks!

**Comment by beingkk** [comment]
> Thanks for that - just before running could you drop the crypt-key in the `nesta-production-config` bucket? Thanks!

Done!

File: innovation_sweet_spots/getters/inputs.py
Comment (jaklinger) [inline innovation_sweet_spots/getters/inputs.py:46]: Just add in the edge case where the file doesn't already locally exist:

```suggestion
    use_cached = use_cached or not os.path.exists(fpath)
    if not use_cached:
```

File: innovation_sweet_spots/getters/inputs.py
Comment (jaklinger) [inline innovation_sweet_spots/getters/inputs.py]: Can ditch the `try /  except` with the above suggested logic

**Review by jaklinger — APPROVED** [review]
Looks good to me - but I've given a couple of extra suggestions to streamline things a little bit more - though they're not crucial (I'll leave it to your conscience :) )

A comment to bear in mind for if this goes into any kind of pipeline is that it might be preferable of have an additional mode which doesn't write to disk at all, and instead returns the raw `projects` without the intermediate file i/o. Obviously for local prototyping this is a bit slow, so no need for now,

File: innovation_sweet_spots/getters/inputs.py
Comment (beingkk) [inline innovation_sweet_spots/getters/inputs.py]:
```suggestion
        projects = json.load(open(fpath, "r"))
        logging.info(f"Loaded in the file {fpath}")
```

File: innovation_sweet_spots/getters/inputs.py
Comment (beingkk) [reply innovation_sweet_spots/getters/inputs.py:46]: I think this should be `use_cached = use_cached and os.path.exists(fpath)` as to use "cache" we definitely need `os.path.exists(fpath)`...

**Comment by beingkk** [comment]
Thanks for the review! I added your comment on pipeline into a new issue, so that I don't forget about it (cf. #17)

**Pull request merged**
